lib/raygun.dart

```diff
@@ -0,0 +1,191 @@
+      throw WarpException(result.error);
+    }
+
+    List<Message> msgs = [];
+    // Iterate over the messages
+    int length = result.data.ref.len;
+    for (int i = 0; i < length; i++) {
+      Message message = Message();
+      Pointer<G_Message> pMsg = result.data.ref.ptr.elementAt(i).value;
+
+      // Message ID
+      message.id = bindings.message_id(pMsg).value.toString();
+      // Conversation ID
+      message.conversationId =
+          bindings.message_conversation_id(pMsg).value.toString();
+      // Sender ID, DID only
+      Pointer<G_SenderId> senderIdPointer = bindings.message_sender_id(pMsg);
+      Pointer<G_DID> senderDid =
+          bindings.sender_id_get_did_key(senderIdPointer);
+      message.senderId = DID(senderDid).toString();
+      // DateTime
+      // TODO: test required
+      // The value from Rust is integer (UTC from Chrono ).
+      // Possibly DateTime.fromMillisecondsSinceEpoch() can be used.
+      message.date = DateTime(bindings.message_date(pMsg).value);
+      // Reactions
+      List<Reaction> reactions = [];
+      Pointer<G_FFIVec_Reaction> pReactions = bindings.message_reactions(pMsg);
+      int reactionsLen = pReactions.ref.len;
+      for (int j = 0; j < reactionsLen; j++) {
+        Reaction reaction = Reaction();
+        Pointer<G_Reaction> pReaction = pReactions.ref.ptr.elementAt(j).value;
+        reaction.emoji = bindings.reaction_emoji(pReaction).toString();
+        Pointer<G_FFIVec_SenderId> reactionSendersId =
+            bindings.reaction_users(pReaction);
+        int reactionSendersIdLen = reactionSendersId.ref.len;
+        for (int k = 0; k < reactionSendersIdLen; k++) {
+          reaction.senderId
+              .add(reactionSendersId.ref.ptr.elementAt(k).value.toString());
+        }
+      }
+      message.reactions = reactions;
+      // Pinned
+      message.pinned = bindings.message_pinned(pMsg) == 0 ? false : true;
+      // Message body
+      Pointer<G_FFIVec_String> lines = bindings.message_lines(pMsg);
+      int lineLen = lines.ref.len;
+      for (int j = 0; j < lineLen; j++) {
+        message.value.add(lines.ref.ptr.value.toString());
+      }
+      msgs.add(message);
+
+      // TODO: Replied - Rust binding is not ready
+      // TODO: Metadata - Rust binding is not ready
+    }
+
+    return msgs;
+  }
+
+  // This method serves for sending and editing.
+  // If there is message sent, Warp will cover the editing part.
+  send(String conversationId, String messageId, Message messages) {
+    // Convert given values to native friendly types
+    Pointer<Int8> _convoId = conversationId.toNativeUtf8().cast<Int8>();
+    Pointer<Int8> _messageId = messageId.toNativeUtf8().cast<Int8>();
+    G_FFIVec_String _messages = messages.value.cast().elementAt(0);
+    int lines = _messages.len;
+
+    bindings.raygun_send(
+        pRaygun, _convoId, _messageId, _messages.ptr.elementAt(0), lines);
+  }
+
+  delete(String conversationId, String messageId) {
```

Comment:
For the sake of sanity, this function could be split into `deleteMessage` and `deleteConversation`. The `bindings.raygun_delete` does have it so that the messageId is optional. If one is supplied the message would be deleted, but if none is provided, the conversation itself would be deleted. We could probably do `String? messageId` (if this is the correct way for optional fields?) but splitting it into two would likely prevent mistakes on dart end. Any opinions on that?

Reply:
I updated for the delete to accept the message as optional parameter.
Then assign the value if the parameter is not null. Otherwise nullptr if the parameter is null.